src/main.c

```diff
@@ -464,10 +466,21 @@ static void async_nb_read_ready(h2o_socket_t *sock, const char *err);
 static void async_nb_on_write_complete(h2o_socket_t *sock, const char *err)
 {
     /* add to the read queue the entry for which we have written the request, and start reading from the socket */
-    async_nb_push(&async_nb.read_queue, async_nb_pop(&async_nb.write_queue));
+    struct async_nb_transaction_t *transaction = async_nb_pop(&async_nb.write_queue);
+    if (transaction->expect_response) {
+        async_nb_push(&async_nb.read_queue, transaction);
+    }
+
     if (!h2o_socket_is_reading(async_nb.sock))
         h2o_socket_read_start(async_nb.sock, async_nb_read_ready);
 
+    if (transaction->allocated) {
+        assert(!h2o_linklist_is_linked(&transaction->link));
+        neverbleed_transaction_dispose(transaction->buf);
+        free(transaction->buf);
+        free(transaction);
+    }
```

Comment:
What is the intended behavior when both `allocated` and `expect_response` are true, or if both are false?

When both of them are true, maybe we see an assertion failure at line 478. But what if both are false? Do we have a memory leak?

I'm wondering if it might be better to have only one boolean flag instead of two, consider that `expect_response == false` means we have to free the buffer and the transaction here.

Reply:
Yeah this can be simplified to a single flag. Addressed in https://github.com/h2o/h2o/pull/3216/commits/1c8e9a17df29bc4cf1217b94cdfe8f0e56878fae